src/api/firebase.js

```diff
@@ -173,3 +173,71 @@ export async function deleteItem() {
 	 * this function must accept!
 	 */
 }
+
+/**
+ * Compare purchase history of list items and sort items by three attributes:
+ * 1) Purchase status, 2) Urgency, and 3) Alphabetically
+ * add a property to each item indicating the urgency of the next purchase
+ * @param {object} data The raw list data to be sorted.
+ */
+export function comparePurchaseUrgency(data) {
+	// Create an empty, nested array for sorted data categories
+	const sortedData = [];
+	for (let i = 0; i < 6; i++) {
+		sortedData.push([]);
+	}
+
+	// Loop through all items and assign urgency based on days until next purchase
+	for (let item of data) {
+		// If the item has been purchased before, return days since last purchase
+		// If the item has never been purchased, return the days since item creation
+		const daysSincePurchase = item.dateLastPurchased
+			? getDaysBetweenDates(item.dateLastPurchased)
+			: getDaysBetweenDates(item.dateCreated);
+
+		// Assign urgency
+		if (daysSincePurchase >= 60) {
+			item.purchaseStatus = 'Inactive';
+			item.urgencyCategory = 4; // Inactive
+		} else {
+			const daysUntilPurchase = item.dateLastPurchased
+				? getDaysBetweenDates(item.dateNextPurchased)
+				: getDaysBetweenDates(undefined, item.dateNextPurchased);
+
+			item.purchaseStatus = 'Active';
+			item.days = daysUntilPurchase;
+
+			if (daysUntilPurchase < 0) {
+				item.urgencyCategory = 0; // Overdue
+			} else if (daysUntilPurchase <= 7) {
+				item.urgencyCategory = 1; // Soon
+			} else if (daysUntilPurchase > 7 && daysUntilPurchase < 30) {
+				item.urgencyCategory = 2; // Kind of Soon
+			} else if (daysUntilPurchase >= 30) {
+				item.urgencyCategory = 3; // Not Soon
+			} else {
+				item.urgencyCategory = 4; // Inactive
+			}
+		}
+	}
```

Comment:
Moving this `assign urgency` logic to its own separate function will help streamline this `comparePurchaseUrgency` function.

The way to think about this is that this `assign urgency` functionality has its own "business logic" that would be able to stand on its own without the comparisons. It's helpful to have a clear separation of concerns when you're writing "business logic".

Reply:
I would also approach this method as a `calculateUrgency` method (just return the urgency) vs an `assignUrgency` method (which would add another property to the item). There are two considerations that make me think that:
1. if you have a method to calculate the urgency (that's not too computationally heavy), there's no real reason to store it in the object itself
2. further downstream when rendering the data, you won't need to grab the item `urgencyCategory` on a per item basis because you've already grouped them by urgency in the 2 dimensional array `sortedData`

Reply:
Hi @Jennifer Batara! Apologies for dragging on this. I had a question that came up; let's see if I can ask it succinctly.. :popcorn:

> I would also approach this method as a calculateUrgency method (just return the urgency) vs an assignUrgency method (which would add another property to the item). There are two considerations that make me think that...

So, if we have an object like `numDaysInEstimate` that is providing a dictionary for our urgencies, are we aiming to return the string eg 'soon', the number of days represented by this string, or the index of the urgency?
This is a big learning for me because I often use array indices and hadn't considered that these are often magic numbers.